cns/healthserver/healthz.go

```diff
@@ -0,0 +1,54 @@
+package healthserver
+
+import (
+	"net/http"
+
+	"github.com/Azure/azure-container-networking/crd/nodenetworkconfig/api/v1alpha"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+)
+
+var schema = runtime.NewScheme()
+
+func init() {
+	utilruntime.Must(v1alpha.AddToScheme(schema))
+}
+
+func NewHealthzHandlerWithChecks() http.Handler {
+	cfg, err := ctrl.GetConfig()
+	if err != nil {
+		panic(err)
+	}
+	cli, err := client.New(cfg, client.Options{
+		Scheme: schema,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	checks := map[string]healthz.Checker{
+		"nnc": func(req *http.Request) error {
+			ctx := req.Context()
+			// we just care that we're allowed to List NNCs so set limit to 1 to minimize
+			// additional load on apiserver
+			if err := cli.List(ctx, &v1alpha.NodeNetworkConfigList{}, &client.ListOptions{
```

Comment:
I wonder if listing NNCs is the best we can do - this may be immediately problematic for cilium nodesubnet where we now run CNS but do not expect NNCs (or even to talk to the API server? @santhoshmprabhu)...

Reply:
anything in the cnsconfig that's reliable to tell us if we need to read NNCs? `CNIConflistScenario         string`? or maybe try and list pods (keeping limit at 1) if ipamv2 is enabled?

Reply:
`ChannelMode==AzureHost` is how we configure nodesubnet. @rbtr, maybe `ChannelMode==CRD || ChannelMode==MultiTenantCRD`  is the right check for NNCs?

Reply:
is there a tl;dr for what these channel modes map to or what they mean?

```
// ChannelMode :- CNS channel modes
const (
	Direct         = "Direct"
	Managed        = "Managed"
	CRD            = "CRD"
	MultiTenantCRD = "MultiTenantCRD"
	AzureHost      = "AzureHost"
)
```

Reply:
@rbtr @santhoshmprabhu lmk what you think now. The nnc checker is only initialized for channelMode `CRD`. I guess we could also include MTCRD but I wanted to keep changes minimal for now (although maybe we should _start_ with MTCRD to keep blast radius smaller).

I'm also open to only returning unhealthy if we get a 401 this is mostly to address the "expired token" issue.